File: 第14课 课件+四种语言模板+代码/【字典树模板】lc208 实现Trie.cpp
class Trie {
public:
    Trie() { root = new Node(); }
    void insert(string word) { find(word, true, true); }    
    bool search(string word) { return find(word, true, false); }    
    bool startsWith(string prefix) { return find(prefix, false, false); }

private:
    struct Node {
        int count;
        unordered_map<char, Node*> child;
        Node(): count(0) {}
    };
    Node* root;

    bool find(const string& s, bool exact_match, bool insert_if_not_exist) {
        Node* curr = root;
        for (char c : s) {
            if (curr->child.find(c) == curr->child.end()) {
                if (!insert_if_not_exist) return false;
                curr->child[c] = new Node();
            }
            curr = curr->child[c];
        }
        if (insert_if_not_exist) curr->count++;
        return exact_match ? curr->count > 0 : true;
    }
};

File: 第14课 课件+四种语言模板+代码/【字典树模板】lc208 实现Trie.py
class Trie:
    def __init__(self):
        self.root = [0, {}]  # [count, child]
    def insert(self, word: str) -> None:
        self.find(word, True, True)
    def search(self, word: str) -> bool:
        return self.find(word, True, False)
    def startsWith(self, prefix: str) -> bool:
        return self.find(prefix, False, False)
    
    def find(self, s, exact_match, insert_if_not_exist):
        curr = self.root
        for ch in s:
            if ch not in curr[1]:
                if not insert_if_not_exist:
                    return False
                curr[1][ch] = [0, {}]
            curr = curr[1][ch]
        if insert_if_not_exist:
            curr[0] += 1
        return curr[0] > 0 if exact_match else True